gh-135953: Profile a module or script with sampling profiler

lkollar opened the pull request:
Add `-m` and `filename` arguments to the sampling profiler to launch the specified Python program in a subprocess and start profiling it. Previously only a PID was accepted, this can now be done by passing `-p PID`.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-135953 -->
* Issue: gh-135953
<!-- /gh-issue-number -->

lkollar committed:
Profile a module or script with sampling profiler

Add `-m` and `filename` arguments to the sampling profiler to launch the
specified Python program in a subprocess and start profiling it.
Previously only a PID was accepted, this can now be done by passing
`-p PID`.

pablogsal commented on Lib/profile/sample.py:
```diff
@@ -712,19 +733,59 @@ def main():
+        process = subprocess.Popen(cmd)
+
+        try:
+            exit_code = process.wait(timeout=0.1)
```
Hummm, why is this? Just in case the process fails immediately?

lkollar replied:
This was to wait until the interpreter in the subprocess has initialised. I replaced it now with a function which retries as a couple of times. Do you think there a better way to handle this?

lkollar committed:
Wait for interpreter to initialize in subprocess

lkollar committed:
Remove unnecessary argument check

These args are already mutually exclusive, but we need to check if at
least on module argument has been passed.

lkollar committed:
Handle the process runs shorter than expected

In this case the subprocess will go into zombie state until we can poll
it. We can simply assume this is the case if it's still detected as
running when we get a ValueError.

lkollar commented on Lib/profile/sample.py:
```diff
@@ -53,6 +57,14 @@ def sample(self, collector, duration_sec=10):
                     break
                 except (RuntimeError, UnicodeDecodeError, MemoryError, OSError):
                     errors += 1
+                except ValueError as e:
```
I wonder if we should add an extra condition here and only do this if we are the ones who spawned the subprocess. Maybe this is fine as-is. What do you think @pablogsal?

pablogsal replied:
Uf, I am not a fun of this as this may be hiding errors in `collector.collect(stack_frames)`. We need some better signal here

lkollar replied:
Actually, as long as `get_stack_trace` can raise a `ProcessNotFoundError` on all platforms, we shouldn't need this as that should capture this situation. We have to fix how this is handled on macOS then as we get a `ValueError` there.

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -53,6 +57,14 @@ def sample(self, collector, duration_sec=10):
+                    if self._is_process_running():
+                        duration_sec = (
+                            current_time - start_time
+                        )
```
Line length is fine here:
```suggestion
                    if self._is_process_running():
                        duration_sec = current_time - start_time
```

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -537,51 +549,95 @@ def _validate_collapsed_format_args(args, parser):
+                print("Waiting for process to start...")
+                time.sleep(STARTUP_RETRY_DELAY_SECONDS)
+            else:
+                raise RuntimeError("Process failed to start after maximum retries")
```
```suggestion
                raise RuntimeError("Process failed to start after maximum retries") from None
```

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -537,51 +549,95 @@ def _validate_collapsed_format_args(args, parser):
 def main():
     # Create the main parser
     parser = argparse.ArgumentParser(
         description=(
```
It could be easier to put this description in a single module-level multi-line string?

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -712,19 +768,42 @@ def main():
-        realtime_stats=args.realtime_stats,
-    )
+    if args.module is not None and not args.module:
+        parser.error("the following arguments are required: module name")
```
Would `argparse.PARSER` work here for `nargs`?

lkollar replied:
We need to avoid picking up any arguments passed to `-m` (e.g. `--foo`) and pass them verbatim to the module. If I'm not mistaken, `PARSER` would attempt to parse these.

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -712,19 +768,42 @@ def main():
+        if args.module:
+            cmd = [sys.executable, "-m", *args.module]
+        else:
+            cmd = [sys.executable, *args.script]
```
```suggestion
        if args.module:
            cmd = (sys.executable, "-m", *args.module)
        else:
            cmd = (sys.executable, *args.script)
```

AA-Turner commented on Lib/profile/sample.py:
```diff
@@ -13,6 +14,9 @@
 
 FREE_THREADED_BUILD = sysconfig.get_config_var("Py_GIL_DISABLED") is not None
+MAX_STARTUP_ATTEMPTS = 5
+STARTUP_RETRY_DELAY_SECONDS = 0.1
```
Perhaps make these private constants (leading underscore)?

lkollar committed:
Improve ReadRemoteMemory error handling on macOS

Improve the return value check to be able to raise a ProcessLookupError
when the remote process is not available.

Mach uses composite error values where higher error values indicate
specific subsystems. We can use the err_get_code function to mask the
higher bits to make our error checking more robust in case the subsystem
bits are set. For example, in some situations if the process is in zombie
state, we can get KERN_NO_SPACE (0x3) but the actual return value is
0x10000003 which indicates a specific subsystem, thus we need to
use err_get_code to extract the error value.

This also improves how KERN_INVALID_ARGUMENT is handled to check whether
we got a generic invalid argument error, or if the process is no longer
accessible.

lkollar committed:
fixup! Handle the process runs shorter than expected

lkollar committed:
fixup! Wait for interpreter to initialize in subprocess

lkollar committed:
fixup! fixup! Wait for interpreter to initialize in subprocess

lkollar committed:
fixup! Improve ReadRemoteMemory error handling on macOS

lkollar committed:
fixup! fixup! Wait for interpreter to initialize in subprocess

pablogsal commented on Python/remote_debug.h:
```diff
@@ -1053,18 +1055,38 @@ _Py_RemoteDebug_ReadRemoteMemory(proc_handle_t *handle, uintptr_t remote_address
+                    "Process %d is no longer accessible (process terminated)",
+                    handle->pid);
+            } else {
+                PyErr_Format(PyExc_PermissionError,
```
This one needs to still be ValueError no? Is not a permission error if we pass an invalid address

lkollar committed:
fixup! fixup! Improve ReadRemoteMemory error handling on macOS

pablogsal committed:
Do not return error without an exception set

pablogsal committed:
Protect against more errors with no exceptions

pablogsal committed:
Attempt to sync processes
